src/ll/gcd.rs

```diff
@@ -0,0 +1,81 @@
+use std::cmp::Ordering;
+
+use ll;
+use ll::limb::Limb;
+
+pub unsafe fn gcd(gp: *mut Limb, mut ap: *mut Limb, mut an: i32, mut bp: *mut Limb, mut bn: i32) -> i32 {
+    assert!(an >= bn);
+
+    let mut gc = 0;
+    while *ap == 0 && !ll::is_zero(ap, an) && *bp == 0 && !ll::is_zero(bp, bn){
+        ap = ap.offset(1);
+        bp = bp.offset(1);
+        an -= 1;
+        bn -= 1;
+        gc += 1;
+    }
+
+    let a_trailing = (*ap).trailing_zeros() as u32;
+    let b_trailing = (*bp).trailing_zeros() as u32;
+
+    let trailing = if a_trailing < b_trailing {
+        a_trailing
+    } else {
+        b_trailing
+    };
+
+    if trailing > 0 {
+        ll::shr(ap, ap, an, trailing);
+        ll::shr(bp, bp, bn, trailing);
+    }
+
+    ll::copy_incr(bp, gp, bn);
+    bp = gp;
+
+    while !ll::is_zero(ap, an) {
+
+        while *ap == 0 && !ll::is_zero(ap, an) {
+            ap = ap.offset(1);
+            an -= 1;
+        }
+
+        let at = (*ap).trailing_zeros() as u32;
+        if at > 0 {
+            ll::shr(ap, ap, an, at);
+        }
+
+        while *bp == 0 && !ll::is_zero(bp, bn) {
+            bp = bp.offset(1);
+            bn -= 1;
+        }
+
+        let bt = (*bp).trailing_zeros() as u32;
+        if bt > 0 {
+            ll::shr(bp, bp, bn, bt);
+        }
+
+        let c = ll::cmp(ap, bp, an);
+
+        // FIXME find better way to compute |ap - bp| / 2
```

Comment:
This looks fine to me. Not sure what better way there would be, other than a custom function for it (which seems silly given how specific it is).